### PR TITLE
OPS-1395 Initial API and hashmap-based static impl

### PR DESCRIPTION
**Pull request by tnn**
@sorenmat @berardino @JesperTerkelsen 

This implementation is not finished, but I would like to get some feedback on the general direction. 

I had the idea of a service-discovery library that could create an interface (API) that can abstract away different implementations, so we don't end up tying ourself to an implementation as we currently have done with Zookeeper / Curator. I couldn't find a generic interface for this and decided this project was small enough to take on now.

My idea was to create a dns-srv (https://www.ietf.org/rfc/rfc2782.txt) ServiceDiscovery implementation and a consul-based (https://www.consul.io/) ServiceRegistry implementation (names are tentative).

### CODE REVIEW

**Comment by sorenmat** [comment]
What is this for again ?

**Comment by berardino** [comment]
could you please give a little more insights on the scope of this project ?
Is it _just_ for service discovery ? if so, what's the strategy are we going for (DNS ?) ?
Did we consider solutions like : curator, eureka ?

**Comment by tnn** [comment]
@sorenmat @berardino Thanks for the awesome quick response. I've updating the pull request description a bit more. I'll try writing a better description in a README.md tomorrow. Please note I created the repository as open and I hope we can keep it as an open source library, so Tradeshift internal talk should probably be on another channel.

**Comment by tnn** [comment]
If you should find some time, could I get you to look at the interfaces - just so it doesn't die completely :)

File: api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+     * <pre>^(?![0-9]+$)(?!-)[a-zA-Z0-9-]{,63}(?&lt;!-)$</pre>.
+     *
+     * @param service RFC1123 compliant name of service to register.
+     * @param node    non-null URI for service to register (self). any path of service or query string is not guranteed
+     *                to be available to hosts to the implementation. Host and protocol most be specified and will be
+     *                lower-cased.
+     * @return future when registration has successfully been applied and is available to others. In the event of
+     * service registry not being available, an ExecutionException will be thrown on the get method of the future.
+     */
+    CompletionStage<Registration> register(String service, URI node);
```
Comment (sorenmat) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java]: Can't you do ?
CompletionStage<Registration> register(@NonNull String service, @NonNull URI node);

File: static/src/main/java/com/tradeshift/servicediscovery/inmemory/InMemoryServiceRegistry.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        return CompletableFuture.completedFuture(registration);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Supplier<Optional<URI>> discover(final String service) {
+        return () -> {
+            final Set<URI> uriSet = fixed.get(service);
```
Comment (sorenmat) [inline static/src/main/java/com/tradeshift/servicediscovery/inmemory/InMemoryServiceRegistry.java]: perhaps clean it up with getOrDefault(service, Optional.empty()) ?

Comment (tnn) [inline static/src/main/java/com/tradeshift/servicediscovery/inmemory/InMemoryServiceRegistry.java]: Yep, I'll try to clean it a bit up, this is not that pretty. I don't like the iterator part either :)

File: api/src/main/java/com/tradeshift/servicediscovery/ServiceDiscovery.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     *
+     * @param service  user-defined service name, Fx. 'xmpp' or 'http'
+     * @return supplier of an non-null optional URI. It is up to the implementation to determine how each invocation of
+     * the Supplier returns a value, but it should be generally assumed that the ServiceDiscovery implementation will
+     * handle priority and load balancing.
+     * @throws NullPointerException     if any argument is null
+     * @throws IllegalArgumentException if protocol is not a valid IANA protocol name or if service is not a valid IANA
+     *                                  service name
+     */
+    Supplier<Optional<URI>> discover(String service);
```
Comment (sorenmat) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceDiscovery.java]: The service seems to be more like a protocol, or am I missing something ?

File: api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java (hunk 1 above)
Comment (tnn) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java]: Is the JSR-305 annotations available? I would prefer not dragging lombok as a dependency.

File: api/src/main/java/com/tradeshift/servicediscovery/ServiceDiscovery.java (hunk 3 above)
Comment (tnn) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceDiscovery.java]: Yeah, I used to have both (protocol, service) and I think my javadoc got outdated.

File: api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java (hunk 1 above)
Comment (sorenmat) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java]: https://blogs.oracle.com/java-platform-group/entry/java_8_s_new_type

Comment (tnn) [inline api/src/main/java/com/tradeshift/servicediscovery/ServiceRegistry.java]: Thanks!

**Pull request closed without merging**